### ISSUE TITLE
libmount: Storing optlist as part of context changes behavior

### ISSUE THREAD

**Issue by laenion** [util-linux/util-linux#2326]
libmount before v2.39 was using the file system options stored in the `libmnt_fs` struct when calling any of the `mnt_context_*` functions. With v2.39 this changed, the options are now stored as part of the context and have to be set explicitly.

**Example**

In the past it was enough to call something like
```
mnt_fs_set_target(mnt_fs, "/mnt");
mnt_cxt = mnt_new_context();
mnt_context_set_fs(mnt_cxt, mnt_fs));
mnt_context_set_mflags(mnt_cxt, some_more_flags));
mnt_context_mount(mnt_cxt);
```
to mount a file system with modified mount options.

Since v2.39 an additional
```
mnt_context_set_options(mnt_cxt, mnt_fs_get_options(mnt_fs));
```
is required to actually use the options from the existing file system. This behavior change was quite unexpected, especially as both API and ABI didn't change.

I'm not sure whether this change was intentional, but as it may be breaking existing applications unnoticed it's probably worth pointing it out.

**Comment by Vogtinator** [comment]
> I'm not sure whether this change was intentional, but as it may be breaking existing applications unnoticed

It definitely broke transactional-update, but that was noticed and at least some code in snapper changed runtime behavior due to this, but not noticably (yet).

**Comment by Vogtinator** [comment]
Demo which just mounts an entry from fstab at `/mnt`:

```
#include <libmount.h>
int main(int argc, char *argv[])
{
        struct libmnt_table *tab = mnt_new_table_from_file("/etc/fstab");
        struct libmnt_fs *fs = mnt_table_find_mountpoint(tab, argv[1], MNT_ITER_FORWARD);
        struct libmnt_context *cxt = mnt_new_context();
        mnt_context_set_fs(cxt, fs);
        mnt_context_set_target(cxt, "/mnt");
        mnt_context_mount(cxt);
}
```

Before 2.39, this applied the options from fstab:

```
30f3e1ab267d:/ # cat /etc/fstab
tmpfs /mnt tmpfs size=0%,noatime 0 0
30f3e1ab267d:/ # ./test /mnt
30f3e1ab267d:/ # findmnt /mnt
TARGET SOURCE FSTYPE OPTIONS
/mnt   tmpfs  tmpfs  rw,noatime,size=0k,uid=1000,gid=100,inode64
```

With 2.39, it silently no longer does:

```
30f3e1ab267d:/ # ./test /mnt
30f3e1ab267d:/ # findmnt /mnt
TARGET SOURCE FSTYPE OPTIONS
/mnt   tmpfs  tmpfs  rw,relatime,uid=1000,gid=100,inode64
```

**Comment by karelzak** [comment]
You're right; this is regression. I'll fix it for v2.39.1.

**Comment by Vogtinator** [comment]
FTR, we backported the fix in our TW package and can confirm that it works and hasn't broken anything (yet).

**Comment by zeha** [comment]
Fixes landed in master and 2.39.1, right? Should this be closed?

**Comment by karelzak** [comment]
Yes, closing.
